src/classes/ALLO_AllocationsRecalculateService.cls

```diff
@@ -0,0 +1,604 @@
+                    allocation.Campaign__c = null;
+
+                    // Copy all Allocations if copyInActiveAllocations is true
+                    if (copyInactiveAllocations) {
+                        listAlloForInsert.add(allocation);
+                    } else {
+                        if (allocation.General_Accounting_Unit__r.Active__c) {
+                            listAlloForInsert.add(allocation);
+                        }
+                    }
+                }
+            }
+
+            Decimal remainder = amount;
+
+            // Update the Allocations amounts and percentages, and if required, non-percentage based Allocations to percentage Allocations
+            for (Allocation__c allocation : listAlloForInsert) {
+                //if this is a multicurrency org, recalculate the amount and set the CurrencyIsoCode to the same one as the parent
+                if (isMultiCurrencyEnabled) {
+                    if (allocation.Amount__c != null) {
+                        allocation.Amount__c = UTIL_CurrencyConversion.convertAmountUsingStandardRates(
+                            (String)allocation.get(UTIL_Currency.CURRENCY_ISO_CODE_FIELD),
+                            (String)targetObj.get(UTIL_Currency.CURRENCY_ISO_CODE_FIELD),
+                            allocation.Amount__c
+                        );
+                    }
+                    allocation.put(UTIL_Currency.CURRENCY_ISO_CODE_FIELD, targetObj.get(UTIL_Currency.CURRENCY_ISO_CODE_FIELD));
+                }
+
+                if (allocation.Percent__c != null) {
+                    allocation.Amount__c = (amount * allocation.Percent__c * .01).setScale(2);
+                }
+
+                //Calculate proportional amount for non-percentage based Payment allocations
+                // convertToPercentages must be true, the Percentage null, sourceAmount cannot be null and either Default Allocations are not enabled or the allocation is not for the Default Allocation
+                if (convertToPercentages && allocation.Percent__c == null & sourceAmount != null) {
+                    Decimal allocationPercent = ((allocation.Amount__c != null) ? allocation.Amount__c : 0) / sourceAmount * 100;
+
+                    // Do not round the percent yet, we want the most accurate Amount value possible
+                    allocation.Amount__c = (amount * allocationPercent * .01).setScale(2);
+
+                    // Round the visible percentage value to 2 decimal places after the amount is calculated
+                    allocation.Percent__c = allocationPercent.setScale(2, System.RoundingMode.HALF_UP);
+                }
+                remainder -= (allocation.Amount__c != null) ? allocation.Amount__c : 0;
+            }
+
+            //if our remainder is not the same sign as the amount, only create percentage based Allocations
+            if ((amount >= 0 && remainder < 0) || (amount < 0 && remainder > 0)) {
```

Comment:
The amount/remainder comparisons throughout this rather large if-block make this hard to read/understand. Can these be assigned into vars and then compared in the if? For example:
```
Boolean isRemainderSameSign = (amount >= 0 && remainder < 0) || (amount < 0 && remainder > 0);

if (isRemainderSameSign) {
...
}
```

And so on with the subsequent ifs.